vmm/src/memory_manager.rs

```diff
@@ -246,50 +261,183 @@ impl BusDevice for MemoryManager {
         ext_regions: Option<Vec<MemoryRegion>>,
         prefault: bool,
     ) -> Result<Arc<Mutex<MemoryManager>>, Error> {
-        // Init guest memory
-        let arch_mem_regions = arch::arch_memory_regions(config.size);
-
-        let ram_regions: Vec<(GuestAddress, usize)> = arch_mem_regions
-            .iter()
-            .filter(|r| r.2 == RegionType::Ram)
-            .map(|r| (r.0, r.1))
-            .collect();
-
         let mut mem_regions = Vec::new();
-        if let Some(ext_regions) = &ext_regions {
-            if ram_regions.len() > ext_regions.len() {
-                return Err(Error::InvalidAmountExternalBackingFiles);
+        let arch_mem_regions: Vec<(GuestAddress, usize, RegionType)>;
+        let use_custom_regions = config.size == 0;
+
+        if !use_custom_regions {
+            if config.regions.is_some() {
+                error!(
+                    "User defined memory regions can't be provided if the \
+                    memory size is not 0"
+                );
+                return Err(Error::InvalidMemoryParameters);
             }
 
-            for region in ext_regions.iter() {
-                mem_regions.push(MemoryManager::create_ram_region(
-                    &Some(region.backing_file.clone()),
-                    region.start_addr,
-                    region.size as usize,
-                    true,
-                    prefault,
-                    false,
-                    false,
-                )?);
+            // Init guest memory
+            arch_mem_regions = arch::arch_memory_regions(config.size);
+
+            let ram_regions: Vec<(GuestAddress, usize)> = arch_mem_regions
+                .iter()
+                .filter(|r| r.2 == RegionType::Ram)
+                .map(|r| (r.0, r.1))
+                .collect();
+
+            if let Some(ext_regions) = &ext_regions {
+                if ram_regions.len() > ext_regions.len() {
+                    return Err(Error::InvalidAmountExternalBackingFiles);
+                }
+
+                for region in ext_regions.iter() {
+                    mem_regions.push(MemoryManager::create_ram_region(
+                        &Some(region.backing_file.clone()),
+                        0,
+                        region.start_addr,
+                        region.size as usize,
+                        true,
+                        prefault,
+                        false,
+                        false,
+                    )?);
+                }
+            } else {
+                for region in ram_regions.iter() {
+                    mem_regions.push(MemoryManager::create_ram_region(
+                        &config.file,
+                        0,
+                        region.0,
+                        region.1,
+                        false,
+                        prefault,
+                        config.shared,
+                        config.hugepages,
+                    )?);
```

Comment:
I think that in order to simplify the code flow, we could internally consider the `--memory size > 0` as a special case of `--memory size=0 --memory-zones size=1G size=4G`.

Essentially, `--memory size=1G` is equivalent to `--memory size=0 --memory-zone size=1G`, i.e we could build a single zone for the `--memory size=1G` case and use the same code path for both `size=0` and `size > 0`.

Reply:
Yes I agree that'd be nicer. I hope you don't mind me adding this as a new commit instead of trying to rebase through existing commits.